transformers/generate/sampler.py

```diff
@@ -0,0 +1,210 @@
+        if self.do_apply_temperature:
+            return logits / self.temperature
+        return logits
+
+    def apply_top_k_filter(self, logits):
+        """ Use the probability distribution of the tokens to determine the set
+        to be sampled from.  Specifically we select the set of size k such that
+        the sum of its items' probabilities is maximum.
+
+        .. Fan, Angela, Mike Lewis, and Yann Dauphin. "Hierarchical neural
+        story generation." arXiv preprint arXiv:1805.04833 (2018).
+        """
+        if self.k > 0:
+            try:
+                indices_to_remove = logits < torch.topk(logits, self.k)[0][..., -1, None]
+                logits[indices_to_remove] = -float("Inf")
+            except RuntimeError:
+                vocabulary_size = logits.size(-1)
+                raise RuntimeError(
+                    "the value of k provided ({}) must be smaller than the size of the vocabulary ({})".format(
+                        self.k, vocabulary_size
+                    )
+                )
+
+        return logits
+
+    def apply_nucleus_filter(self, logits):
+        """ Use the probability distribution of the tokens to determine the set
+        to be sampled from. Specifically, choose the smallest set such that the
+        sum of its items' probabilities is greater than a number p in [0,1].
+
+        .. Holtzman, Ari, et al. "The curious case of neural text
+           degeneration." arXiv preprint arXiv:1904.09751 (2019).
+        """
+        if self.p > 0:
+            sorted_logits, sorted_indices = torch.sort(logits, descending=True)
+            sorted_probabilities = F.softmax(sorted_logits, dim=-1)
+            cumulative_probabilities = torch.cumsum(sorted_probabilities, dim=-1)
+
+            # Remove tokens with cumulative probability above the threshold,
+            # but keep the first token above the threshold.
+            sorted_indices_to_remove = cumulative_probabilities > self.p
+            sorted_indices_to_remove[..., 1:] = sorted_indices_to_remove[..., :-1].clone()
+            sorted_indices_to_remove[..., 0] = 0
+
+            # scatter sorted tensors to original indexing
+            indices_to_remove = sorted_indices_to_remove.scatter(
+                dim=-1, index=sorted_indices, src=sorted_indices_to_remove
+            )
+            logits[indices_to_remove] = -float("Inf")
+
+        return logits
+
+    def sample_one_token(self, logits):
+        if self.do_apply_temperature:
+            return torch.multinomial(F.softmax(logits, dim=-1), num_samples=1)
+        return torch.argmax(logits, dim=-1).unsqueeze(-1)
+
+
+class SamplerSingleStack(Sampler):
+    """ Generic sampler for single-stack models.
+    """
+
+    def __init__(self, model, config, device):
+        self.model = model
+        super(SamplerSingleStack, self).__init__(config, device)
+
+    def generate_sequence(self, length=1, prompt_ids=[], **model_kwargs):
+        prompt_ids = torch.tensor(
```

Comment:
We should take as inputs tensors and return tensors in this method so it can be called in the `decode` methods of the `PretrainedModel` class.